### PR TITLE
DashboardGrid: Add support to filter panels using variable

### LINKED ISSUE #74891: Panel Filtering in Dashboard
Hi there. We moving over dashboards from an in-house legacy system to our Grafana instance. Something that's a bit unique about our current system is that users have created very large dashboards, with sometimes thousands of panels. This is possible to navigate because our in-house system provides a panel filter where a user can specify filtering regex and only display only those panel's whose title matches said regex. Finally, the regex is included in the URL so that a filtered view can be shared.

We realize this is an impractical paradigm and the obvious solution would be to force users to sub-divide or reorganize dashboards prior to moving to Grafana. However, due to the constraints of our migration to Grafana the only option is to do an initial copy-paste, followed by a dashboard refactor at our user's leisure.

I intend to submit a PR to add this panel regex filter feature and place it under a feature flag. However, I'd still like to ask for feedback to ensure the implementation is elegant and architecturally sound. With that I'd like to ask:

* Would it be more strategic to place this functionality into a new variable type? This way the [URL support](https://grafana.com/docs/grafana/latest/dashboards/build-dashboards/create-dashboard-url-variables/) would come for free
* If the variable route is not recommended, can you recommend a place to start? (e.g. New component, wiring to URL component, etc). I want to make sure I structure this neatly and not just duct-tape the feature in
* Could this feature be made more generic so that others can benefit? Our use case requires regex filtering but maybe we can make different types available (e.g. prefix filtering)

Thanks again and please forgive my shallow understanding of Grafana's internals.

**Why is this needed**:
- We require legacy dashboards to be moved as is
- These dashboards sometimes have thousands of panels
- We need to provide our users with a 1:1 panel filtering experience

**What would you like to be added**:
- A way to filter panels in a dashboard based on user inputed regex
- The regex should be encoded in the URL so that the state of the dashboard can be shared

**Who is this feature for?**
- For a large migration away from a legacy in-house solution to Grafana

### PR DESCRIPTION
**Pull request by jacobzelek**
**What is this feature?**

Users can add a special variable to their dashboards that will allow them to filter which panels are being displayed.

**Why do we need this feature?**

Only a small subset of users may need this feature. It's particular focus is for migrations that involve dashboards with hundreds, if not thousands of dashboards that require a fast way of drilling down.

**Who is this feature for?**

The feature was originally proposed to assists users during a migration from a legacy metrics platform. The details are mentioned in #74891.

**Which issue(s) does this PR fix?**:

Fixes #74891 

**Special notes for your reviewer:**

Please check that:
- [X] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by s0lesurviv0r** [comment]
@dprokop Something went wrong in the other PR. I think it was when I rebased it. I took a diff and applied it to this branch instead.

**Comment by s0lesurviv0r** [comment]
Screenshot from 2023-10-25 00-59-21

I'm worried I broke more than one thing today :sweat_smile:

**Comment by torkelo** [comment]
It does not work for me, when I change the textbox variable value and hit enter nothing happens. Have to do a full page reload for the filtering to work.

File: public/app/features/dashboard/dashgrid/DashboardGrid.tsx [hunk 1]
```diff
@@ -37,21 +46,71 @@ export class DashboardGrid extends PureComponent<Props> {
...
+            this.setPanelFilter(variable.query);
+          }
+          break;
+        }
+      }
+    }
+
     this.eventSubs.add(dashboard.events.subscribe(DashboardPanelsChangedEvent, this.triggerForceUpdate));
+
+    if (config.featureToggles.panelFilterVariable) {
```
Comment (torkelo) [inline public/app/features/dashboard/dashgrid/DashboardGrid.tsx:72]: can we combine these to if statements into one?

File: public/app/features/dashboard/dashgrid/DashboardGrid.tsx [hunk 2]
```diff
@@ -78,13 +137,27 @@ export class DashboardGrid extends PureComponent<Props> {
...
+      if (!panelFilter) {
+        layout.push(panelPos);
+      } else {
+        if (panelFilter.test(panel.title)) {
+          panelPos.isResizable = false;
+          panelPos.isDraggable = false;
+          panelPos.x = (count % 2) * GRID_COLUMN_COUNT;
+          panelPos.y = Math.floor(count / 2);
+          layout.push(panelPos);
+          count++;
```
Comment (torkelo) [inline public/app/features/dashboard/dashgrid/DashboardGrid.tsx:149]: how do you restore the position when the filtering is removed?

File: public/app/features/dashboard/dashgrid/DashboardGrid.tsx (hunk 1 above)
Comment (s0lesurviv0r) [reply public/app/features/dashboard/dashgrid/DashboardGrid.tsx:72]: No problem

File: public/app/features/dashboard/dashgrid/DashboardGrid.tsx (hunk 2 above)
Comment (s0lesurviv0r) [reply public/app/features/dashboard/dashgrid/DashboardGrid.tsx:149]: `panelPos` is created every time the layout is recomputed which would happen on the `panelFilter` changing

**Comment by s0lesurviv0r** [comment]
Apologies @torkelo I think I missed a few files from the other branch. I'll recheck these and update the PR

**Comment by s0lesurviv0r** [comment]
Hey @dprokop this is failing because I removed the feature flag from documentation. The test case that failures assumes every feature flag will be documented in `docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md`. Is there a way of keeping the feature flag but not documenting it? or should we just document it anyways?

**Review by natellium — APPROVED** [review]
From a behaviour perspective this looks good. It behaves a bit oddly in a couple of scenarios:

- Rows are part of the search as well, making a grid a bit odd if they match the search
- It doesn't work for panels that have a template variable in the title. The search seems to happen before the interpolation, so you cannot search by the value of that variable

That said, these are probably edge cases given your scenario and could be looked into later if needed.

**Comment by s0lesurviv0r** [comment]
Hi @dprokop and @natellium anything else you need from my side to get this merged?

**Comment by dprokop** [comment]
@s0lesurviv0r - nope, nothing, I'm just trying to make the build green and I'll merge :)

**Review by dprokop — APPROVED** [review]

**Pull request merged**